sys/riscv/exception.S

```diff
@@ -1,13 +1,300 @@
 #include <riscv/asm.h>
+#include <riscv/riscvreg.h>
 
-ENTRY(cpu_exception_handler)
-	# Not implemented!
-	j .
-END(cpu_exception_handler)
+#include "assym.h"
 
-ENTRY(user_exc_leave)
-	# Not implemented!
-	j .
-END(user_exc_leave)
+.macro save_reg reg, offset
+	REG_S	\reg, (\offset)(sp)
+.endm
+
+.macro save_reg_cfi reg, offset
+	save_reg \reg, \offset
+	.cfi_rel_offset	\reg, \offset
+.endm
+
+.macro save_csr csr, offset, tmp
+	csrr	\tmp, \csr
+	save_reg \tmp, \offset
+.endm
+
+.macro load_reg reg, offset
+	REG_L	\reg, (\offset)(sp)
+.endm
+
+.macro load_csr csr, offset, tmp
+	load_reg \tmp, \offset
+	csrw	\csr, \tmp
+.endm
+
+/* 
+ * MODE:
+ *  - 0: we came from userspace
+ *  - 1: we came from kernelspace
+ */
+.macro save_ctx mode
+.if \mode == 1
+	PTR_ADDI	sp, sp, -CTX_SIZE
+.endif
+
+	save_reg_cfi ra, CTX_RA
+
+.if \mode == 0
+	/* Load kernel's global pointer. */
+	save_reg_cfi gp, CTX_GP
+	LOAD_GP()
+
+	/* Load pcpu pointer. */
+	save_reg_cfi tp, CTX_TP
+	PTR_LA	tp, _pcpu_data
+.endif
+
+	save_reg_cfi t0, CTX_T0
+	save_reg_cfi t1, CTX_T1
+	save_reg_cfi t2, CTX_T2
+	save_reg_cfi t3, CTX_T3
+	save_reg_cfi t4, CTX_T4
+	save_reg_cfi t5, CTX_T5
+	/* XXX: we don't provide the CFI directive for $t6 as
+	 * GDB from toolchain 1.5.3 has faulty frame info implementation
+	 * which confuses this register with PC. */
```

Comment:
What with latest gdb?

Reply:
I tired it with the following toolchain: (GCC 12.1.0, Binutils 2.38, GDB 12.1). The same issue still occurs. I updated the comment to explicitly specify what configuration was used.

Reply:
For my curiosity - could you check llvm build with lldb?